File: 96.py
class Solution(object):
    def numTrees(self, n):
        """
        :type n: int
        :rtype: int
        """
        result = [0 for _ in range(n+1)]
        result[0] = 1
        result[1] = 1
        if n<= 1: return result[n]
        for i in range(2, n+1):
            for j in range(i):
                result[i] += result[j] * result[i-j-1]
        return result[n]

File: 96.cpp
class Solution {
public:
    int numTrees(int n) {
        vector<int> table(n+1,0);
        table[0] = 1;
        for (int size = 1; size <= n; ++size)
            for (int pivot = 1; pivot <= size; ++ pivot){
                int left = table[pivot-1];
                int right = table[size - pivot];
                table[size] += left * right;
            }
        return table[n];
    }
};
